### PR TITLE
Lua spawn functions multistate

### PR DESCRIPTION
**Pull request by dannuic**
Create lua spawn and ground item tables in every thread's state by lazily constructing it and updating it if the script ever uses it.

### CODE REVIEW

File: src/plugins/lua/LuaThread.cpp [hunk 1]
```diff
@@ -657,4 +658,80 @@ void LuaThread::AssociateTopLevelObject(const MQTopLevelObject* tlo)
...
+	{
+		sol::table spawn_table = m_globalState.create_named_table("__spawns");
+
+		if (pSpawnManager != nullptr)
+		{
+			auto spawn = pSpawnManager->FirstSpawn;
+
+			while (spawn != nullptr)
+			{
+				WriteChatf("Adding spawn %s", spawn->Name);
```
Comment (brainiac) [inline src/plugins/lua/LuaThread.cpp:674]: need to remove the debug messages

File: src/plugins/lua/LuaThread.cpp [hunk 2]
```diff
@@ -657,4 +658,80 @@ void LuaThread::AssociateTopLevelObject(const MQTopLevelObject* tlo)
...
+	}
+}
+
+void LuaThread::AddSpawn(eqlib::PlayerClient* spawn)
+{
+	if (m_globalState["__spawns"] != sol::nil)
+	{
+		WriteChatf("Adding spawn %s", spawn->Name);
+		m_globalState["__spawns"][spawn->SpawnID] = bindings::lua_MQTypeVar(datatypes::pSpawnType->MakeTypeVar(spawn));
+	}
```
Comment (brainiac) [inline src/plugins/lua/LuaThread.cpp:688]: These functions are doing the `__spawns` lookup twice. We should save the result from the first so we can cut these down in half (they happen in bulk)

Comment (dannuic) [reply src/plugins/lua/LuaThread.cpp]: good call -- I did that, but would it be better if I stored these tables as members on LuaThread directly? I'd have to update it when it gets set from nil, but that should work as well, no?

Comment (brainiac) [reply src/plugins/lua/LuaThread.cpp]: yeah that would be even better if it works as expected.

Comment (dannuic) [reply src/plugins/lua/LuaThread.cpp]: I added and tested this, with an extra guard against trying to write/access lua tables that don't exist anymore due to a finished or bad lua state

**Review by brainiac — APPROVED** [review]

**Comment by dannuic** [comment]
merged with 22649fcb0a9af2991593999ab6c3ec3df3b911af

**Pull request closed without merging**